Sources/TuistCore/Extensions/AbsolutePath+Extras.swift

```diff
@@ -35,9 +35,24 @@ extension AbsolutePath {
 
     /// Returns a function to sorten files for Xcode projects.
     /// - Returns: Sortening function.
-    public static func xcodeSortener() -> ((AbsolutePath, AbsolutePath) -> Bool) {
+    public static func xcodeSortener(fileHandler: FileHandling) -> ((AbsolutePath, AbsolutePath) -> Bool) {
         return { lhs, rhs in
-            lhs.components.count < rhs.components.count || lhs.asString < rhs.asString
+            let decompose: (inout [AbsolutePath], String) -> Void = { $0.append($0.last!.appending(component: $1)) }
+            let lhsPathsSet = Set(lhs.components.dropFirst().reduce(into: [AbsolutePath("/")], decompose))
+            let rhsPathsSet = Set(rhs.components.dropFirst().reduce(into: [AbsolutePath("/")], decompose))
+
+            let commonPaths = lhsPathsSet.intersection(rhsPathsSet)
+
+            guard let lhsPathFirst = lhsPathsSet.subtracting(commonPaths).sorted().first else { return false }
+            guard let rhsPathFirst = rhsPathsSet.subtracting(commonPaths).sorted().first else { return false }
+
+            if fileHandler.isFolder(lhsPathFirst), !fileHandler.isFolder(rhsPathFirst) {
+                return false
+            } else if !fileHandler.isFolder(lhsPathFirst), fileHandler.isFolder(rhsPathFirst) {
+                return true
+            } else {
+                return lhsPathFirst < rhsPathFirst
+            }
```

Comment:
Is it possible to write a test for this?

Reply:
I wonder if we could achieve a similar result without needing to access the file system - what was the scenario that was failing previously? - as @ollieatkinson points out, would be handy to have a few test cases to highlight the scenarios.

Reply:
It tries to sort the files and folders in a group first by its type (file or folder) and then by its name. There must be another way to achieve the same result but I couldn't come up with anything better. Any ideas?

Reply:
Also, giving you more context on this. I found this out because I tried to write the same project several times and I noticed that the order of the children inside a group was changing for every write.

Reply:
the closest thing I can come up with is https://github.com/tuist/tuist/pull/257/files#diff-ec115f5bdb34f2bbf56b4df47f965cebR15 however this approach would require a recursive sort of all children as a post step.

Doing it an advance as you have already done is probably easier to reason about